src/systemd/sd-bus-vtable.h

```diff
@@ -64,10 +64,12 @@ struct sd_bus_vtable {
                         const char *result;
                         sd_bus_message_handler_t handler;
                         size_t offset;
+                        const char *names;
```

Comment:
so, add_object_vtable_internal() checks whether the `element_size` field of the first vtable entry equals our assumption of the entry size. This check will start to fail when old code uses our new library. This needs to be catered for. i.e. the check there should be tweaked to allow both the old size of the structure and the new size. i.e. look for this line:

```c
assert_return(vtable[0].x.start.element_size == sizeof(struct sd_bus_vtable), -EINVAL);
```

it should be changed to something like this:

```c
/* Current structure size */
#define VTABLE_ELEMENT_SIZE_NEW sizeof(struct sd_bus_vtable)

/* Structure size up to v241 */
#define VTABLE_ELEMENT_SIZE_OLD \ 
        MAX3(offsetof(struct sd_bus_vtable, x.method_old) + sizeof(*((struct sd_bus_vtable*) NULL)->method_old)),
             offsetof(struct sd_bus_vtable, x.signal_old) + sizeof(*((struct sd_bus_vtable*) NULL)->signal_old)),
             offsetof(struct sd_bus_vtable, x.properties) + sizeof(*((struct sd_bus_vtable*) NULL)->properties)))


assert_return(IN_SET(vtable[0].x.start.element_size, VTALE_ELEMENT_SIZE_OLD, VTABLE_ELEMENT_SIZE_NEW), -EINVAL);
```

for this to work the `method_old` and `singal_old` union fields would have to be retained that contain the precise definitions from the old systemd versions. We'd only use them for the sizeof() above to work, and would not have any other use for them.

Reply:
I prefer to save the entire structure, it is cleaner as procedure and for the resulting code.
Backward compatibility comes with a cost, but it is unavoidable.

Reply:
the problem appeared to be deeper than that.
every time that we loop through the vtable elements with v++, this is converted to the size of the **current** structure, so that after the _SD_BUS_VTABLE_START the structure is misaligned.
It has been necessary to adjust this in both add_object_vtable_internal() and bus_slot_disconnect(). I hope there are no other occurrences of such loops. Could you verify this? Here it is very important to review.

The idea of the test case was really good indeed, because then the test failed and I could see the issue!

Reply:
yeah, there might be various places that need updating. But this should be easy enough to hide behind a macro of some form.